### PR TITLE
Feature/9041 quantity estimation form ondemand

### PR DESCRIPTION
**Pull request by nehajain008**
Story [AB#9041](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/9041)
Task [AB#11727](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/11727)
Task [AB#11728](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/11728)

### CODE REVIEW

**Comment by nehajain008** [comment]
Story AB#9041
Task AB#11729
Task AB#11730
Task AB#11732
Task AB#11735
Task AB#11736
Task AB#11731

**Comment by nehajain008** [comment]
Story AB#9158
Task AB#11750
Task AB#11751
Task AB#11752
Task AB#11753
Task AB#11754
Task AB#11755
Task AB#11756
Task AB#11757

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/ondemand/manifest.json [hunk 1]
```diff
@@ -12,7 +13,7 @@
     },
     "quantity": {
       "id": "quantity",
-      "mainAdvice": "How many do you think you'll need?",
+      "mainAdvice": "How many do you think you’ll need?",
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/flat/ondemand/manifest.json:16]: I don't think we need to change the apostrophe character.

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/patient/ui-tests/general.ui.test.js [hunk 2]
```diff
@@ -37,7 +37,6 @@ const catalogueSolutionIdInSession = 'solution-1';
 const deliveryDateInSession = '2020-10-10';
 const recipientsInSession = JSON.stringify(recipients);
 const selectedRecipientsInSession = JSON.stringify(selectedRecipients);
-
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/flat/patient/ui-tests/general.ui.test.js:40]: I don't think we need to remove this blank line.

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+    harness.request(context, ($) => {
+      const backLink = $('[data-test-id="go-back-link"]');
+      expect(backLink.length).toEqual(1);
+      expect(backLink.text().trim()).toEqual('Go back');
+      expect($(backLink).find('a').attr('href'))
+        .toEqual('/organisation/order-1/catalogue-solutions/select/solution/price/recipients/date');
+    });
+  }));
+  it('should render the "Continue" button', componentTester(setup, (harness) => {
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js:25]: We should add a blank line between the the two tests.

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+      continueButtonText: 'Continue',
+    };
+
+    harness.request(context, ($) => {
+      const button = $('[data-test-id="continue-button"] button');
+      expect(button.length).toEqual(1);
+      expect(button.text().trim()).toEqual(context.continueButtonText);
+    });
+  }));
+  it('should render error summary with correct error text and hrefs if there are errors', componentTester(setup, (harness) => {
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js:36]: We should add a blank line between the the two tests.

File: app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/manifest.json [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    "DeliveryDateRequired": "Enter a planned delivery date",
+    "DeliveryDateInvalid": "Enter a planned delivery date in a valid format",
+    "DeliveryDateOutsideDeliveryWindow": "Planned delivery date must be within 42 months from the commencement date for this Call-off Agreement",
+    "DeliveryDateNotReal": "Planned delivery date must be a real date",
+    "DeliveryDateDayRequired": "Planned delivery date must include a day",
+    "DeliveryDateMonthRequired": "Planned delivery date must include a month",
+    "DeliveryDateYearRequired": "Planned delivery date must include a year",
+    "DeliveryDateYearLength": "Year must be four numbers, for example 2020"
+  }
+}
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/manifest.json:84]: We don't need to support tiered prices yet so we can remove the code that relates to tiered pricing. However, keep a copy (either in a stash and/or a copy of the files) as we may be able to make use of it later.

File: app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/general.ui.test.js [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    .selectText(monthInput).pressKey('delete')
+    .expect(monthInput.hasClass('nhsuk-input--error')).notOk()
+    .selectText(yearInput).pressKey('delete')
+    .expect(yearInput.hasClass('nhsuk-input--error')).notOk()
+    .click(saveButton);
+
+  await t
+    .click(errorSummary.find('li a').nth(1))
+    .expect(getLocation()).eql(`${pageUrl}#deliveryDate`);
+});
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/general.ui.test.js:361]: We don't need to support tiered prices yet so we can remove the code that relates to tiered pricing. However, keep a copy (either in a stash and/or a copy of the files) as we may be able to make use of it later.

File: app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/withoutSavedData.ui.test.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    .click(saveButton);
+
+  await t
+    .expect(errorSummary.exists).ok()
+    .expect(errorSummary.find('li a').count).eql(1)
+    .expect(await extractInnerText(errorSummary.find('li a').nth(0))).eql(content.errorMessages.QuantityGreaterThanZero)
+
+    .expect(solutionTableError.exists).ok()
+    .expect(await extractInnerText(solutionTableError)).contains(content.errorMessages.QuantityGreaterThanZero);
+});
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/withoutSavedData.ui.test.js:144]: We don't need to support tiered prices yet so we can remove the code that relates to tiered pricing. However, keep a copy (either in a stash and/or a copy of the files) as we may be able to make use of it later.

File: app/pages/sections/order-items/catalogue-solutions/routes.js [hunk 8]
```diff
@@ -53,15 +53,13 @@ export const catalogueSolutionsRoutes = (authProvider, addContext, sessionManage
   router.get('/:orderItemId', authProvider.authorise({ claim: 'ordering' }), withCatch(logger, authProvider, async (req, res) => {
     const { orderId, orderItemId } = req.params;
     const accessToken = extractAccessToken({ req, tokenType: 'access' });
-
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/routes.js:56]: We don't need to delete this blank line.

File: app/pages/sections/order-items/catalogue-solutions/routes.js [hunk 9]
```diff
@@ -53,15 +53,13 @@ export const catalogueSolutionsRoutes = (authProvider, addContext, sessionManage
...
     const pageData = await getOrderItemPageDataBulk({
       req,
       sessionManager,
       accessToken,
       orderId,
       orderItemId,
     });
-
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/routes.js:64]: We don't need to delete this blank line.

File: app/pages/sections/order-items/catalogue-solutions/select/routes.js [hunk 10]
```diff
@@ -294,5 +311,75 @@ export const catalogueSolutionsSelectRoutes = (authProvider, addContext, session
...
+      itemName,
+      formData,
+    });
+    logger.info(`navigating to order ${orderId} catalogue-solutions ${provisioningType} form`);
+    if (priceType === 'flat' && provisioningType === 'patient') {
+      return res.redirect(`${config.baseUrl}/organisation/${orderId}/catalogue-solutions/neworderitem`);
+    }
+    return res.render(`pages/sections/order-items/catalogue-solutions/order-item/${priceType}/template.njk`, addContext({ context, user: req.user, csrfToken: req.csrfToken() }));
+  }));
+  router.post('/solution/price/:priceType/:provisioningType', authProvider.authorise({ claim: 'ordering' }), withCatch(logger, authProvider, async (req, res) => {
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/select/routes.js:345]: Blank line between the route handler functions.

File: app/pages/sections/order-items/catalogue-solutions/select/routes.test.js [hunk 11]
```diff
@@ -614,30 +615,23 @@ describe('catalogue-solutions select routes', () => {
           expect(res.text.includes('data-test-id="error-title"')).toEqual(false);
         });
     });
+  });
+  describe('GET /organisation/:orderId/catalogue-solutions/select/solution/price/flat/ondemand', () => {
```
Comment (bjss-chris) [inline app/pages/sections/order-items/catalogue-solutions/select/routes.test.js:619]: Blank line between function definitions.

**Review by bjss-chris — CHANGES_REQUESTED** [review]
You've added some code for tiered pricing but we don't need that to meet the acceptance criteria for the current stories. Therefore, I've recommended that you remove that code but keep it stashed for potential use in the future.

I'm not sure if it's as a result of your changes but the call-off ID is appearing in lower case on a number of pages, for example:

image

However, it should be 'C010000-01'. @PhysicsJon, is this an existing issue?

Also, after saving per patient prices and returning to the page the price list sizes are now empty:

image

I don't think this was the case before. However, unless it is an easy fix we might want to open a separate bug ticket for that to avoid this PR getting any larger.

File: app/helpers/controllers/validateOrderItemTypeForm.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
+/* eslint-disable no-restricted-globals */
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/helpers/controllers/validateOrderItemTypeForm.js]: removed

File: app/helpers/controllers/validateOrderItemTypeForm.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    type: selectedPrice.type,
+  });
+
+  if (selectedPriceManifest.questions.quantity) {
+    if (!data.quantity || data.quantity.trim().length === 0) {
+      errors.push({
+        field: 'Quantity',
+        id: 'QuantityRequired',
+      });
+    } else if (isNaN(data.quantity)) {
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/helpers/controllers/validateOrderItemTypeForm.js]: replaced as suggested

File: app/helpers/controllers/validateOrderItemTypeForm.test.js [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+  itemUnit: {
+    name: 'consultation',
+    description: 'per consultation',
+  },
+  price: 0.1,
+};
+
+const orderItemType = 'some-order-item-type';
+
+describe('validateOrderItemForm', () => {
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/helpers/controllers/validateOrderItemTypeForm.test.js:22]: added

File: app/pages/sections/order-items/catalogue-solutions/order-item/contextCreator.js [hunk 15]
```diff
@@ -19,7 +19,13 @@ export const getContext = ({
...
     ? ((errorMap.quantity || {}).errorMessages || [''])
       .concat((errorMap.deliveryDate || {}).errorMessages) : undefined;
-
+  let newbackLinkRef = '';
+  if (selectedPrice.provisioningType === 'Patient') {
+    newbackLinkRef = `${baseUrl}/organisation/${orderId}/catalogue-solutions/select/solution/price/recipients/date`;
+  } else {
+    newbackLinkRef = `${baseUrl}/organisation/${orderId}/catalogue-solutions/select/solution/price/${selectedPrice.type}/${selectedPrice.provisioningType}`
+      .toLowerCase();
+  }
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/contextCreator.js]: changed as suggested

File: app/pages/sections/order-items/catalogue-solutions/order-item/contextCreator.test.js [hunk 16]
```diff
@@ -32,53 +32,54 @@ describe('catalogue-solutions order-item contextCreator', () => {
       const context = getContext({
         commonManifest,
         orderItemId: 'neworderitem',
+        selectedPrice: { type: 'flat', provisioningType: 'ondemand' },
       });
-      expect(context.backLinkHref).toEqual('/order/organisation/undefined/catalogue-solutions/select/solution/price/recipients/date');
+      expect(context.backLinkHref).toEqual('/order/organisation/undefined/catalogue-solutions/select/solution/price/flat/ondemand');
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/contextCreator.test.js:48]: added another test case

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/commonManifest.json [hunk 17]
```diff
@@ -0,0 +1,6 @@
+{
+    "backLinkText": "Go back",
+    "title": "Quantity of",
+    "continueButtonText": "Continue"
+  }
+
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/flat/commonManifest.json]: deleted extra space

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/declarative/manifest.json [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+      "DeliveryDateInvalid": "Enter a planned delivery date in a valid format",
+      "DeliveryDateOutsideDeliveryWindow": "Planned delivery date must be within 42 months from the commencement date for this Call-off Agreement",
+      "DeliveryDateNotReal": "Planned delivery date must be a real date",
+      "DeliveryDateDayRequired": "Planned delivery date must include a day",
+      "DeliveryDateMonthRequired": "Planned delivery date must include a month",
+      "DeliveryDateYearRequired": "Planned delivery date must include a year",
+      "DeliveryDateYearLength": "Year must be four numbers, for example 2020"
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/flat/declarative/manifest.json]: deleted extra space

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/ondemand/manifest.json (hunk 1 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/flat/ondemand/manifest.json]: reverted to original

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/patient/ui-tests/general.ui.test.js (hunk 2 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/flat/patient/ui-tests/general.ui.test.js]: added blank line

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js (hunk 3 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js]: added blank line

File: app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js (hunk 4 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/flat/template.test.js]: added blank line

File: app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/manifest.json (hunk 5 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/manifest.json]: removed tiered folder  also kept a copy of the files.

File: app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/general.ui.test.js (hunk 6 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/general.ui.test.js]: Removed tiered folder also kept a copy of the files.

File: app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/withoutSavedData.ui.test.js (hunk 7 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/order-item/tiered/patient/ui-tests/withoutSavedData.ui.test.js]: Removed tiered folder also kept a copy of the files.

File: app/pages/sections/order-items/catalogue-solutions/routes.js (hunk 8 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/routes.js]: Added blank line.

File: app/pages/sections/order-items/catalogue-solutions/routes.js (hunk 9 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/routes.js]: Added blank line.

File: app/pages/sections/order-items/catalogue-solutions/select/routes.js (hunk 10 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/select/routes.js]: Added blank line.

File: app/pages/sections/order-items/catalogue-solutions/select/routes.test.js (hunk 11 above)
Comment (nehajain008) [reply app/pages/sections/order-items/catalogue-solutions/select/routes.test.js]: Added blank line.

**Comment by nehajain008** [comment]
> 
> 
> You've added some code for tiered pricing but we don't need that to meet the acceptance criteria for the current stories. Therefore, I've recommended that you remove that code but keep it stashed for potential use in the future.
> 
> I'm not sure if it's as a result of your changes but the call-off ID is appearing in lower case on a number of pages, for example:
> 
> image
> 
> However, it should be 'C010000-01'. @PhysicsJon, is this an existing issue?
> 
> Also, after saving per patient prices and returning to the page the price list sizes are now empty:
> 
> image
> 
> I don't think this was the case before. However, unless it is an easy fix we might want to open a separate bug ticket for that to avoid this PR getting any larger.

>Fixed above mentioned issues.

**Review by bjss-chris — APPROVED** [review]

**Pull request merged**